src/SqlBinding/SqlAsyncCollector.cs

```diff
@@ -90,55 +92,284 @@ public async Task FlushAsync(CancellationToken cancellationToken = default)
+                Dictionary<string, string> columnDataMatchPOCO = new Dictionary<string, string>(columnNamesFromPOCO.Count(), StringComparer.OrdinalIgnoreCase);
+                foreach (string pocoColumn in columnNamesFromPOCO)
+                {
+                    columnDataMatchPOCO.Add(pocoColumn, columnDefinitionsFromSQL[pocoColumn]);
+                }
+
+                // 4. Query SQL for table Primary Keys
+                var primaryKeys = new List<string>();
+                try
+                {
+                    await sqlConnection.OpenAsync();
+                    SqlCommand cmd = new SqlCommand(GetPrimaryKeysQuery(schema, tableName), sqlConnection);
+                    using (SqlDataReader rdr = await cmd.ExecuteReaderAsync())
                     {
-                        // Commands have already been generated, so we just need to attach them to the dataAdapter. No need to 
-                        // discover the table schema
-                        SqlCommand insertCommand = builder.GetInsertCommand();
-                        insertCommand.Connection = connection;
-                        insertCommand.Transaction = transaction;
-                        dataAdapter.InsertCommand = insertCommand;
-                        SqlCommand updateCommand = builder.GetUpdateCommand();
-                        updateCommand.Connection = connection;
-                        updateCommand.Transaction = transaction;
-                        dataAdapter.UpdateCommand = updateCommand;
+                        while (rdr.Read())
+                        {
+                            primaryKeys.Add(rdr[ColumnName].ToString());
+                        }
                     }
-                    dataAdapter.UpdateBatchSize = 1000;
-                    dataAdapter.Update(dataSet, tableName);
-                    await transaction.CommitAsync();
                 }
+                catch (Exception ex)
+                {
+                    // Throw a custom error so that it's easier to decipher.
+                    string message = $"Encountered exception while retrieving primary keys for table '{tableName}.' Cannot generate upsert command without them.";
+                    throw new InvalidOperationException(message, ex);
+                }
+                finally
+                {
+                    await sqlConnection.CloseAsync();
+                }
+
+                if (!primaryKeys.Any())
+                {
+                    string message = $"Did not retrieve any primary keys for '{tableName}.' Cannot generate upsert command without them.";
+                    throw new InvalidOperationException(message);
+                }
+
+                // Match column names to POCO field/property objects
+                IEnumerable<MemberInfo> primaryKeyFields = typeof(T).GetMembers().Where(f => primaryKeys.Contains(f.Name, StringComparer.OrdinalIgnoreCase));
```

Comment:
We might want to consider logging the results of the schema lookup we do to make debugging easier if something doesn't behave the way we expect.